pkg/clusteragent/clusterchecks/dispatcher_ksm.go

```diff
@@ -0,0 +1,183 @@
+		return false
+	}
+
+	// Check if this KSM check should be sharded
+	if !d.ksmSharding.ShouldShardKSMCheck(config) {
+		// KSM check but not suitable for sharding (e.g., only one resource group)
+		log.Infof("KSM check %s not suitable for sharding, using normal scheduling", config.Digest())
+		return false
+	}
+
+	// Check if this config was already sharded (to avoid re-sharding on every Schedule call)
+	if d.isAlreadySharded(config) {
+		log.Debugf("KSM config %s already sharded, skipping re-sharding", config.Digest())
+		return true // Prevent the original config from being scheduled on top of shards
+	}
+
+	// Check if we have any cluster check runners available
+	// Note: Even with 0 or 1 runners, we still create shards
+	// Rebalancing will handle distribution as runners come online
+	runnerCount, _ := d.store.CountNodeTypes()
+
+	if runnerCount == 0 {
+		log.Warnf("KSM sharding is enabled but no cluster check runners (CLC runners) are currently available.")
+	}
+
+	// Create sharded configs based on resource groups
+	// Always do sharding (pods, nodes, others) regardless of current runner count
+	// Rebalancing will automatically distribute shards optimally as runners scale up/down
+	shardedConfigs, err := d.ksmSharding.CreateShardedKSMConfigs(config, runnerCount)
+	if err != nil {
+		log.Warnf("Failed to create resource-sharded KSM configs: %v, falling back to normal scheduling", err)
+		return false
+	}
+
+	log.Infof("Created %d resource-sharded KSM configs", len(shardedConfigs))
+
+	// Schedule resource-sharded configs using dispatcher's logic
+	// Shards are created and tracked regardless of current runner availability
+	// They will be picked up by runners as they come online (via dangling configs)
+	shardedDigests := make([]string, 0, len(shardedConfigs))
+
+	for _, cfg := range shardedConfigs {
+		patchedCfg, err := d.patchConfiguration(cfg)
+		if err != nil {
+			log.Warnf("Cannot patch resource-sharded KSM config %s: %s", cfg.Digest(), err)
+			continue
+		}
+
+		// Add the shard (will go to dangling if no runners available)
+		d.add(patchedCfg)
+		shardedDigests = append(shardedDigests, patchedCfg.Digest())
+	}
+
+	if len(shardedDigests) == 0 {
+		log.Errorf("KSM sharding enabled but failed to create any shards - falling back to normal scheduling")
+		return false
+	}
+
+	log.Infof("Successfully sharded KSM check into %d resource-grouped checks", len(shardedDigests))
+
+	// Store that we've sharded this config to avoid re-sharding
+	d.markAsSharded(config, shardedDigests)
+
+	return true
+}
+
+// isAlreadySharded checks if a KSM config was already sharded
+// Thread-safe: protected by ksmShardingMutex
+func (d *dispatcher) isAlreadySharded(config integration.Config) bool {
+	d.ksmShardingMutex.Lock()
+	defer d.ksmShardingMutex.Unlock()
+
+	if d.ksmShardedConfig.Name != "" && d.ksmShardedConfig.Digest() == config.Digest() {
```

Comment:
This can be simplified: `return d.ksmShardedConfig.Name != "" && d.ksmShardedConfig.Digest() == config.Digest()`